uniffi_bindgen/src/interface/types/mod.rs

```diff
@@ -66,55 +71,12 @@ pub enum Type {
-    /// able to name a particular type in order to e.g. call a helper function that is specific
-    /// to that type. We support this by defining a naming convention where each type gets a
-    /// unique canonical name, constructed recursively from the names of its component types (if any).
-    pub fn canonical_name(&self) -> String {
-        match self {
-            // Builtin primitive types, with plain old names.
-            Type::Int8 => "i8".into(),
-            Type::UInt8 => "u8".into(),
-            Type::Int16 => "i16".into(),
-            Type::UInt16 => "u16".into(),
-            Type::Int32 => "i32".into(),
-            Type::UInt32 => "u32".into(),
-            Type::Int64 => "i64".into(),
-            Type::UInt64 => "u64".into(),
-            Type::Float32 => "f32".into(),
-            Type::Float64 => "f64".into(),
-            Type::String => "string".into(),
-            Type::Boolean => "bool".into(),
-            // API defined types.
-            // Note that these all get unique names, and the parser ensures that the names do not
-            // conflict with a builtin type. We add a prefix to the name to guard against pathological
-            // cases like a record named `SequenceRecord` interfering with `sequence<Record>`.
-            // However, types that support importing all end up with the same prefix of "Type", so
-            // that the import handling code knows how to find the remote reference.
-            Type::Object(nm) => format!("Type{}", nm),
-            Type::Error(nm) => format!("Type{}", nm),
-            Type::Enum(nm) => format!("Type{}", nm),
-            Type::Record(nm) => format!("Type{}", nm),
-            Type::CallbackInterface(nm) => format!("CallbackInterface{}", nm),
-            Type::Timestamp => "Timestamp".into(),
-            Type::Duration => "Duration".into(),
-            // Recursive types.
-            // These add a prefix to the name of the underlying type.
-            // The component API definition cannot give names to recursive types, so as long as the
-            // prefixes we add here are all unique amongst themselves, then we have no chance of
-            // acccidentally generating name collisions.
-            Type::Optional(t) => format!("Optional{}", t.canonical_name()),
-            Type::Sequence(t) => format!("Sequence{}", t.canonical_name()),
-            Type::Map(t) => format!("Map{}", t.canonical_name()),
-            // A type that exists externally.
-            Type::External { name, .. } | Type::Wrapped { name, .. } => format!("Type{}", name),
-        }
-    }
+    // To add a new type:
+    //   - Add a variant to this enum
+    //   - In `dispatch.rs` add a type handler struct and a case in `dispatch_type_function!`
+    //   - Implement `NewCodeType` for the type handler
+    //   - For each language, implement `[Language]CodeType` for the type handler (Note: currently
+    //     only supported on Kotlin)
```

Comment:
What's the failure mode if someone doesn't do this last step? Will the compile give them enough clues to figure out what they missed?

Reply:
It's not the best, but not that mysterious.  The compiler tells you that you're missing a method and maybe you need to implement a trait.  `KotlinCodeType` is listed, but so are other traits.  I think it would be better if the `dispatch_type_function` macro was updated to use `<HandlerObj as $trait_name>`, but that makes the macro more complicated.  I think each match needs to become a block rather than a single expression.  Maybe it's worth it though.  What do you think?